### PR TITLE
ensure a zipped DAG can locate the modules it provides 

### LINKED ISSUE #10616: modules provided and imported by packaged DAGs can get mixed up
**Apache Airflow version**: 1.10.1, 1.10.10, code identical in 1.10.12 (reproduced with unit tests)

**Kubernetes version (if you are using kubernetes)** (use `kubectl version`): N/A

**Environment**:

- **Cloud provider or hardware configuration**: on-prem
- **OS** (e.g. from /etc/os-release):  CentOS 7
- **Kernel** (e.g. `uname -a`): Linux 3.10.0-957.1.3.el7.x86_64
- **Install tools**:
- **Others**:

**What happened**:

Our DAG require some proprietary support library, which evolves from time to time. 
The typical structure of our dag zips ("dagpacks") is the following:

dag-a.zip:
* dag-a.py
* support_lib/
   * __init__.py
   * constants.py
   * functions.py

dag-b.zip:
* dag-b.py
* support_lib/
   * __init__.py
   * constants.py
   * functions.py

While over time, all dagpacks will tend to use the same version of the support lib, the release cycles of various DAGs are decorrelated. 
What happens is that sometimes, and apparently randomly, dag-a will receive dag-b.zip's support_lib instead of its own.

After careful analysis of the logs and the UI, it may even happen that the situation visible in the UI can be different in the scheduler process (i.e. you can have dag-a receiving dag-b's support lib in the Web UI, while the scheduler might behave as if dag-b received dag-a's support library instead !)

**What you expected to happen**:
When a zipped DAG imports a module it provides within the same zip file, it receives that module and nothing else, _ever_. Distinct DAG zips can coexist with distinct versions.

**How to reproduce it**:
Unit test provided in the forthcoming PR

As it happens, the following structure *seems* fine:
dag-a.zip:
* dag-a.py
* support_lib.py

dag-b.zip:
* dag-b.py
* support_lib.py

but migrating back to this would require us to merge back all submodules of our support library into a single file.

**Anything else we need to know**:

This problem seems to happen 100% of the time on our systems. It got unnoticed at first, as the scope of changes within the support_lib was initially timid, but it no longer is.

### PR DESCRIPTION
**Pull request by cchepelov**
closes: #10616

This PR provides a fix for the issue where DAG packages (zips) which provides possibly distinct versions of a support library might receive the wrong version from a different DAG

~~To facilitate the review, it isn't yet squashed as requested; there are four commits~~: _(edit: after rebasing it to master, keeping the four commits wasn't feasible anymore. )_

1. Unit tests that demonstrate the issue (and fail)
2. a refactoring of DagBag.process_file to break up that function into more manageable pieces 
3. a first-level resolution of the issue (effectively, every module loaded by a DAG within a zip is immediately taken off the global namespace, so that the next DAG pack can try again), which still fails in case a DAG zip attempts to provide a different version of an otherwise globally-supplied module (e.g. an updated tzdb or something like that)
4. a correction to the previous, ensuring zipped DAG get loaded in a clean and neutral environment
5. an improvement on the previous, in order to avoid unloading modules between files of the same DAG package (they will be intended to share dependencies if they have any)

(Another issue that is fixed in passing is that previously, the sys.path would grow indefinitely each time the DagBag loads or reloads a zipped DAG)

### CODE REVIEW

**Comment by cchepelov** [comment]
(CI failed on an error which doesn't seem related... Huh?:

`     Running setup.py install for pyodbc: finished with status 'error'
    Complete output from command /usr/local/bin/python -u -c "import setuptools, tokenize;__file__='/tmp/pip-install-3b6bnkqc/pyodbc/setup.py';f=getattr(tokenize, 'open', open)(__file__);code=f.read().replace('\r\n', '\n');f.close();exec(compile(code, __file__, 'exec'))" install --record /tmp/pip-record-3iim2fgk/install-record.txt --single-version-externally-managed --compile:
    running install
    running build
    running build_ext
    building 'pyodbc' extension
    creating build
    creating build/temp.linux-x86_64-3.8
    creating build/temp.linux-x86_64-3.8/src
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DPYODBC_VERSION=4.0.30 -I/usr/local/include/python3.8 -c src/buffer.cpp -o build/temp.linux-x86_64-3.8/src/buffer.o -Wno-write-strings
    In file included from src/buffer.cpp:12:
    src/pyodbc.h:56:10: fatal error: sql.h: No such file or directory
     #include <sql.h>
              ^~~~~~~
    compilation terminated.
    error: command 'gcc' failed with exit status 1
    `

**Comment by mik-laj** [comment]
Does this problem also occur with the master branch? We do not accept changes to branch v1-10-*, if this change has not been merged into the master branch. After merging into the master branch, if possible, the change is cherry-picj to branch v1-10 by the release maintainer.

**Comment by cchepelov** [comment]
Thank you for the prompt reply @mik-laj! 

Looking at master:

- the code has been refactored for legibility already, so there won't be any cherry-picking in either direction
- the core part of the dag zip-loading code doesn't appear materially changed between 1.10.12 and master, so there is a high likelihood the issue affects master as well.

I'll port the PR forward and push it again.

**Comment by amr-noureldin** [comment]
@cchepelov do you have an ETA on when this fix will be ready? We are also impacted by the exact same issue, and I am wondering if we shall wait for the next Airflow release, or try to manually integrate your changes into our fork.

Thanks anyway for your efforts here!

**Comment by mik-laj** [comment]
if you can, please integrate this patch yourself.  This PR target to Airflow 1.10 and we are focusing on Airflow 2,0 development. There is a long way to go before this change could be released in the 1.10 release,

First, we need to prepare an PR for Airflow 2.0.  Once it gets merged we can work on a change for Airflow 1.10, but we have a lot of differences between releases, so we would have to prepare many other dependent changes. we have too much work on Airflow 2.0 to spend time on this bug. I would be happy to help if you would like to start working on a change for Airflow 2.0.

On other hand, we are getting closer to 2.0 and we also need to have some "nice feature" in 2.0 to get more people exited to migrate to it.

**Comment by cchepelov** [comment]
My apologies, I’ve been sidetracked and could not yet finish the patch against the 2.0 branch (playing whack-a-mole with additional ’sys.path leaks’), although I’d reckon I’m 95% there.

I hope to publish it tomorrow. 

Kamil is right you probably ought to take the patch as I originally sent it and attempt to integrate it into your system if that works for you.

**Comment by cchepelov** [comment]
(why those conflicts? T_T I thought I had just rebased...)

**Comment by mik-laj** [comment]
@cchepelov  If you have made a change to your PR but you haven't notified anyone and you don't have a reviewer, it is unlikely that the change will be noticed.
See: 
https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow
If you have made a change to your PR, you must at least write a comment in your PR. Ideally, you should find a reviewer on Slack. It will be helpful to ask for a reviewer on the #development channel.

**Comment by cchepelov** [comment]
indeed @mik-laj ; although I still need a couple cleanups before it's ready for review. 

Slack, ugh...

File: airflow/models/dagbag.py
Comment (ashb) [inline airflow/models/dagbag.py:388]: You're going to have to explain this in a lot more detail. -- there are many many more system modules than this :)

**Review by ashb — CHANGES_REQUESTED** [review]
Does this problem affect just the webserver? If so, as I mentioned in #10616 this fix isn't needed with 2.0 making DAG serialization mandatory (as the webserver never even _reads_ the DAG code anymore.

**Comment by cchepelov** [comment]
it affects task execution as well. 

In certain cases, I've witnessed the (1.10.1) webserver happily pick up the correct module, and then some task executor instances appear fine, and then *somehow* another task execution instance loads things in a different order and then the wrong module version kicks in.

File: airflow/models/dagbag.py
Comment (cchepelov) [reply airflow/models/dagbag.py:388]: fair; there are many more criteria than just being named one of these five modules.

**Comment by ashb** [comment]
Module loading code is the same, but the execution environment of tasks is _very_ different between the versions.

**Comment by ashb** [comment]
Basically my concern here is that this is a lot of code, and not all python modules are safe to just delete from sys.module (specifically things with compiled components might not be happy with this) -- and the following two constraints should already hold in 1.10.12 with DAG serialization enabled, or in 2.0.0 always:

- When parsing a dag file, it is done in a "single-use" process that parses one file, then exits.
- When executing a task, the same caveat should apply - the dag is never loaded in a long-running/persistent python process.

So my worry is this is a lot of code and "messing about" in python internals that serves no end in 2.0.0 and onwards.

**Comment by abhilash1in** [comment]
@ashb Are you saying this issue will not exist in 2.0.0?

**Comment by ashb** [comment]
> > serves no end in 2.0.0 and onwards.
> 
> @ashb Are you saying this issue will not exist in 2.0.0?

That is what I think, yes.

**Pull request closed without merging**
